openwisp_controller/config/whois/services.py

```diff
@@ -0,0 +1,103 @@
+    # Particularly useful in scenarios where many instances of the class are created.
+    # Reference: https://wiki.python.org/moin/UsingSlots
+    __slots__ = ("device", "org_settings")
+
+    _CACHE_KEY_KEY = "whois_info:{ip_address}"
+
+    def __init__(self, device):
+        self.device = device
+        self.org_settings = device._get_organization__config_settings()
+
+    def _need_whois_lookup(self, initial_ip, new_ip):
+        """
+        Returns True if the WhoIs lookup is needed.
+        This is used to determine if the WhoIs lookup should be triggered
+        when the device is saved.
+
+        The lookup is triggered if:
+            - WhoIs is enabled in the organization settings.
+            - The new IP address is not None.
+            - The WhoIs information is not already present or the initial IP is
+              different from the new IP.
+            - The new IP address is a global (public) IP address.
+        """
+        whois_info = self.get_whois_info()
+        return (
+            getattr(self.org_settings, "whois_enabled", WHOIS_ENABLED)
+            and new_ip
+            and (not whois_info or initial_ip != new_ip)
+            and ip_address(new_ip).is_global
+        )
+
+    def trigger_whois_lookup(self):
+        """
+        Trigger WhoIs lookup based on the conditions of `_need_whois_lookup`.
+        Task is triggered on commit to ensure redundant data is not created.
+        """
+        if self._need_whois_lookup(self.device._initial_last_ip, self.device.last_ip):
+            transaction.on_commit(
+                lambda: fetch_whois_details.delay(
+                    device_pk=self.device.pk,
+                    initial_ip_address=self.device._initial_last_ip,
+                    new_ip_address=self.device.last_ip,
+                )
+            )
+
+    def get_whois_info(self):
+        """
+        Returns WhoIsInfo for the device if WhoIs is enabled and IP is valid public ip.
+        Results are cached to avoid repeated lookups for the same IP address.
+        """
+        WhoIsInfo = load_model("config", "WhoIsInfo")
+
+        if not getattr(self.org_settings, "whois_enabled", WHOIS_ENABLED):
+            return None
+
+        ip = self.device.last_ip
+        if not ip or not ip_address(ip).is_global:
+            return None
+
+        key = self._CACHE_KEY_KEY.format(ip_address=ip)
+        whois_info = cache.get(key)
+        if whois_info is not None:
+            return whois_info
+
+        whois_info = WhoIsInfo.objects.filter(ip_address=ip).first()
+        if whois_info:
+            cache.set(key, whois_info, timeout=60 * 60 * 24)  # Cache for 24 hours
+        return whois_info
+
+    # The function is staticmethod because it does not depend on the instance
+    # of the class and deletes the WhoIs record for a given IP address.
+    @staticmethod
+    def delete_whois_record(ip_address):
```

Comment:
Where are we using this function?

Reply:
When deleting record along with device in `device_whois_info_delete_handler` and in `fetch_whois_details` when deleting old whois record (refering to commit : https://github.com/openwisp/openwisp-controller/pull/1054/commits/0b1bef4adb0f02c1688dba3633a15396e6df58d8)